### PR TITLE
cudagraphs dynamo backend

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80566

This backend handles cases where the preexisting cuda graphs
implementation from dynamo is unsound/has errors.

Requires this functorch bug fix: https://github.com/pytorch/functorch/pull/935

Signed-off-by: Edward Z. Yang <ezyangfb.com>

### CODE REVIEW

**Comment by ezyang** [comment]
cc @ngimel @wconstab

**Comment by ezyang** [comment]
cc @SherlockNoMad

File: test/test_dynamo_cudagraphs.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+except ImportError:
+    TEST_DYNAMO = False
+
+TEST_CUDA = torch.cuda.is_available()
+
+if not TEST_CUDA or not TEST_DYNAMO:
+    print('CUDA or dynamo not available, skipping tests', file=sys.stderr)
+    TestCase = object  # noqa: F811
+
+torchdynamo.config.verify_correctness = True
```
Comment (jansel) [inline test/test_dynamo_cudagraphs.py:20]: Move to a decorator (perhaps unittest.mock.patch), import should not have side effects.

File: test/test_dynamo_cudagraphs.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        if self.graph is not None:
+            assert len(args) == len(self.static_inputs)
+            for dst, src in zip(self.static_inputs, args):
+                dst.copy_(src)
+            self.graph.replay()
+            return tree_map(cloner, self.static_outputs)
+
+        elif self.warmed_up:
+            # record
+            self.static_inputs = [x.clone() for x in args]
```
Comment (jansel) [inline test/test_dynamo_cudagraphs.py:49]: use `cloner()`?

File: torch/fx/passes/backends/cudagraphs.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        n.meta['fake_result'] = result
+        return result
+
+    def propagate(self, *args):
+        # TODO: this is not compositional
+        with FakeTensorMode.push() as mode:
+            fake_args = [mode.from_tensor(a) for a in args]
+            return super().run(*fake_args)
+
+def partition_cudagraphs(gm, inputs):
```
Comment (jansel) [inline torch/fx/passes/backends/cudagraphs.py:44]: The other thing cudagraphs doesn't support is input mutation, which we should be able to detect.

File: test/test_dynamo_cudagraphs.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            self.graph = torch.cuda.CUDAGraph()
+            with torch.cuda.graph(self.graph):
+                # TODO: this is wrong but required due to _wrapped_call
+                self.static_outputs = super(type(gm), gm).__call__(*self.static_inputs)
+            return tree_map(cloner, self.static_outputs)
+
+        else:
+            # warmup
```
Comment (wconstab) [inline test/test_dynamo_cudagraphs.py:57]: what does warmup actually do for graph recording?

our docs suggest to warm up for "a few iterations" (in their example they use 3) but idk what for.   
https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/#api-example

File: torch/fx/passes/backends/cudagraphs.py (hunk 3 above)
Comment (ezyang) [reply torch/fx/passes/backends/cudagraphs.py:44]: Good catch, should be easy enough to do with functionalization

File: test/test_dynamo_cudagraphs.py (hunk 4 above)
Comment (ngimel) [reply test/test_dynamo_cudagraphs.py:57]: I think it's ddp, not nccl, but I'm not sure. Cudnn benchmarking throws an error if someone tries to capture it.

Comment (ezyang) [reply test/test_dynamo_cudagraphs.py:57]: oops you're right

File: test/test_dynamo_cudagraphs.py (hunk 2 above)
Comment (jansel) [reply test/test_dynamo_cudagraphs.py:49]: Yes.  The same should be true for return values too.

**Comment by ezyang** [comment]
Substantial refactor of many parts of the code, I also found a bunch of bugs which I will be filing issues for tomorrow

**Comment by ezyang** [comment]
The screamer bug is somehow cuda graphs, during recording, just loses input mutations. Crazy.

**Comment by ngimel** [comment]
I don't think that's a bug, graph recording don't execute the kernels, it only records the succession of launches.

File: test/test_dynamo_cudagraphs.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+                    mut_arg = True
+            if mut_arg:
+                self.mutated_inputs |= self.inputs[StorageWeakRef(argument.storage())]
+        return super().call_function(target, args, kwargs)
+
+    def __call__(self, *args):
+        super().run(*args)
+        return self.mutated_inputs
+
+class ProxyTensorInterpreter(torch.fx.Interpreter):
```
Comment (ezyang) [inline test/test_dynamo_cudagraphs.py:129]: TODO: Move this into a shared area. @Chillee I wanted you to look over this

File: test/test_dynamo_cudagraphs.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    def call_module(self, target, args, kwargs):
+        assert not kwargs
+        # Don't trace the module, but do run the module to get the correct
+        # out result
+        out = super().call_module(target, tree_map(unwrap_elem, args), kwargs)
+        submod = self.module.get_submodule(target)
+        mutated_inputs = FindInputMutations(submod)(*map(unwrap_elem, args))
+        # smh the module didn't get transferred wut
+        self.new_module.add_submodule(target, CudaGraphModule(submod, mutated_inputs))
+        return wrap_output(out, torch.fx.Proxy(self.new_graph.call_module(target, tree_map(unwrap_proxy_node, args), tree_map(unwrap_proxy_node, kwargs)), self.tracer))
```
Comment (ezyang) [inline test/test_dynamo_cudagraphs.py:174]: This is kind of blegh

File: test/test_dynamo_cudagraphs.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+def cudagraphs(model, inputs):
+    model = partition_cudagraphs(model, inputs)
+
+    # Your interpreter
+    t = ApplyCudaGraphs(model)
+    with FakeTensorMode.push() as mode:
+        t.run(*map(mode.from_tensor, inputs))
+    model = t.new_module
+    model.recompile()
```
Comment (ezyang) [inline test/test_dynamo_cudagraphs.py:184]: This is a lot wordier than it should be

File: torch/fx/passes/backends/cudagraphs.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    def run_node(self, n: Node):
+        result = super().run_node(n)
+        n.meta['fake_result'] = result
+        return result
+
+    def propagate(self, *args):
+        # TODO: this is not compositional
+        with FakeTensorMode.push() as mode:
+            fake_args = [mode.from_tensor(a) for a in args]
+            return super().run(*fake_args)
```
Comment (ezyang) [inline torch/fx/passes/backends/cudagraphs.py:42]: This should be moved somewhere else

**Comment by ezyang** [comment]
Cc @Chillee

File: test/test_dynamo_cudagraphs.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    def __init__(self, gm):
+        super().__init__(gm)
+        self.inputs = defaultdict(set)
+        self.input_idx = 0
+        self.mutated_inputs = set()
+
+    def placeholder(self, target, args, kwargs):
+        r = super().placeholder(target, args, kwargs)
+        # NB: inputs could be aliased
+        self.inputs[StorageWeakRef(r.storage())].add(self.input_idx)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_dynamo_cudagraphs.py:109]: Huh that is unexpected...

[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_dynamo_cudagraphs.py:109]: Are the storage objects distinct?

[the remark this answers is not in the document]
Comment (bdhirsh) [reply test/test_dynamo_cudagraphs.py:109]: yep - at least what I was seeing was:
```
b = a.view(-1)

a.storage() == b.storage() # False
StorageWeakRef(a.storage()) == StorageWeakRef(b.storage()) # False
StorageWeakRef(a.storage()).cdata == StorageWeakRef(b.storage()).cdata # True
```

[the remark this answers is not in the document]
Comment (ngimel) [reply test/test_dynamo_cudagraphs.py:109]: I'm seeing StorageWeakRef being fine (maybe it won't be for subclasses?)
```
In [7]: b = a.view(-1)
   ...: 
   ...: print(a.storage() == b.storage(),
   ...: a.storage().data_ptr() == b.storage().data_ptr(), # False
   ...: StorageWeakRef(a.storage()) == StorageWeakRef(b.storage()), # False
   ...: StorageWeakRef(a.storage()).cdata == StorageWeakRef(b.storage()).cdata # True)
   ...: )
False True True True
```
`data_ptr` is also pretty robust.
How would this work for e.g. sparse tensors that don't have `.storage()`?

File: torch/fx/passes/backends/cudagraphs.py (hunk 8 above)
Comment (ezyang) [reply torch/fx/passes/backends/cudagraphs.py:42]: This pass isn't sound; because we save fake tensors directly on nodes, if a graph has metadata changing operation like `resize_` it will mutate the fake tensor

**Comment by ezyang** [comment]
I'd like to merge this into master, can I get some reviews?

**Review by ngimel — APPROVED** [review]
LGTM

File: torch/fx/passes/fake_tensor_prop.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+__all__ = ['FakeTensorProp']
+
+@compatibility(is_backward_compatible=False)
+class FakeTensorProp(torch.fx.Interpreter):
+    """
+    Execute an FX graph Node-by-Node and record a fake tensor representing
+    the metadata for the node.  Unlike ShapeProp, (1) this propagation
+    is cheap--it does the propagation with meta tensors which do not actually
+    store data, and (2) the fake tensors have much more fine grained
```
Comment (wconstab) [inline torch/fx/passes/fake_tensor_prop.py:14]: finish sentence?

File: torch/fx/passes/backends/cudagraphs.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+def partition_cudagraphs(gm, inputs):
+    """
+    Partition an FX graph into sub-GraphModules that can be validly run under
+    CUDA graphs.  For a subgraph to be runnable under CUDA, all of the operations
+    must involve CUDA tensors only/
+    """
+
+    FakeTensorProp(gm).propagate(*inputs)
+    supported_ops = CudaGraphsSupport()
+    # TODO: single node partition may be wrong due to the pessimization
```
Comment (wconstab) [inline torch/fx/passes/backends/cudagraphs.py:48]: what would it mean if you disallowed single-node partitioning?  Is the 'good' case for single-node that there are 2 big cuda graphs surrounding a single-node cpu island?  but there is no prevention for having many small islands or even cuda-graph islands?

**Review by wconstab — APPROVED** [review]
lgtm. do you have benchmark results though?

File: torch/fx/passes/backends/cudagraphs.py (hunk 11 above)
Comment (ezyang) [reply torch/fx/passes/backends/cudagraphs.py:48]: A single node partition would be us trying to CUDA graph a single operator call on CUDA arguments, like at::add. This could still be profitable if the kernel has a lot of CPU overhead (which is the case with, e.g., TensorIterator), but this is counterbalanced by the need to call into and out of the CUDA graph.

The single node partition does NOT refer to CPU islands, those can be arbitrarily small; we're just talking about the size of the fusion groups

**Comment by ezyang** [comment]
On densenet on quadro's it runs 30% faster; other models it makes things slower. I was intending to run on A100 but I haven't setup the benchmarking setup yet

**Comment by ngimel** [comment]
Making things slower is due to extra copies introduced for each graphed segment?

**Comment by ezyang** [comment]
That is what I would assume, but I haven't run an ablation study yet

File: torch/fx/passes/backends/cudagraphs.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
+import torch
+from torch.fx.passes.infra.partitioner import CapabilityBasedPartitioner
+from torch.fx.passes.operator_support import OperatorSupport
+from torch.fx.passes.tools_common import CALLABLE_NODE_OPS
+from torch.fx.passes.fake_tensor_prop import FakeTensorProp
+from torch.utils._pytree import tree_map
```
Comment (msaroufim) [inline torch/fx/passes/backends/cudagraphs.py:6]: curious how come `_pytree` is considered an internal only

File: test/test_dynamo_cudagraphs.py [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+N_ITERS = 5
+
+
+class TestDynamoCudaGraphs(TestCase):
+    @patch_all()
+    def test_basic(self):
+        def model(x, y):
+            return (x + y) * y
+
+        with torchdynamo.optimize(aot_autograd_cudagraphs):
```
Comment (msaroufim) [inline test/test_dynamo_cudagraphs.py:75]: n00b q: Is the optimization unrolled outside of the `with` scope? If you called `torchdynamo.optimize()` in a loop would the result be the same as calling it once?

File: test/test_dynamo_cudagraphs.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+            return a * 3
+
+        with torchdynamo.optimize(aot_autograd_cudagraphs):
+            for i in range(N_ITERS):
+                x = torch.randn(3, device="cuda", requires_grad=True)
+                y = torch.randn((), device="cpu")
+                loss = model(x, y).sum()
+                loss.backward()
+
+    @patch("functorch._src.config.use_functionalize", True)
```
Comment (msaroufim) [inline test/test_dynamo_cudagraphs.py:109]: noob q: I've seen the terminology functionalize a few times and my understanding is you remove stateful operations to ship to compilers that can't represent aliasing. Is that the majority of compilers? Some compilers we really care about?

File: torch/cuda/_dynamo_graphs.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+from torch.multiprocessing.reductions import StorageWeakRef
+from torch.utils._pytree import tree_map
+import torchdynamo  # type: ignore[import]
+from torchdynamo.optimizations.training import AOTAutogradStrategy  # type: ignore[import]
+
+import operator
+from collections import defaultdict
+from typing import Set
+
+# TODO: maybe this should live in torchdynamo instead
```
Comment (msaroufim) [inline torch/cuda/_dynamo_graphs.py:14]: noob comment: It does seem kinda cool that you can register `AOTAutogradStrategy` without changing `dynamo` or `pytorch` code

File: torch/cuda/_dynamo_graphs.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+    # these are all None or all filled
+    graph = None
+    static_inputs = None
+    static_outputs = None
+
+    # NB: we override __call__ as we don't need any nn.Module machinery
+    # and to reduce overhead
+    def __call__(self, *args):
+        # TODO: once we've recorded here, we'd like to replace the __call__
```
Comment (msaroufim) [inline torch/cuda/_dynamo_graphs.py:44]: Is this comment the general PT 2.0 strategy where you do graph surgery by swapping graph nodes for compiled code?

File: torch/cuda/_dynamo_graphs.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
+import torch
+from torch.fx import GraphModule
+from torch.nn import Module
+from torch.fx.passes.backends.cudagraphs import partition_cudagraphs
```
Comment (msaroufim) [inline torch/cuda/_dynamo_graphs.py:4]: what does partitioning actually mean here? Looking at the function seems to indicate that it removes all non CUDA operations from a graph

File: torch/cuda/_dynamo_graphs.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    # these are all None or all filled
+    graph = None
+    static_inputs = None
+    static_outputs = None
+
+    # NB: we override __call__ as we don't need any nn.Module machinery
+    # and to reduce overhead
+    def __call__(self, *args):
+        # TODO: once we've recorded here, we'd like to replace the __call__
+        # implementation with compiled bytecode that copies into static, replays
```
Comment (msaroufim) [inline torch/cuda/_dynamo_graphs.py:45]: is graph capture what NVIDIA would call what we call tracing?

File: torch/fx/passes/backends/cudagraphs.py [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+import operator
+
+class CudaGraphsSupport(OperatorSupport):
```
Comment (msaroufim) [inline torch/fx/passes/backends/cudagraphs.py:10]: Ok I think I get what is happening at a high level

1. You check to see if an fx graph works with your CUDA graph and if not then partition it and remove mutation
2. You take your fx graph and then stitch in CUDA graph modules where appropriate. For each CUDA graph you warm it up with by opening a CUDA stream and and "record and replay" (trace?) the operations
3. Finally you register a new AOTAutogradStrategy that does all of the above automatically if you use it within a `with optimize()` scope

File: torch/fx/passes/backends/cudagraphs.py (hunk 12 above)
Comment (ezyang) [reply torch/fx/passes/backends/cudagraphs.py:6]: We should make it public; we weren't sure if we had stabilized the API enough when we originally added it. cc @zou3519

File: test/test_dynamo_cudagraphs.py (hunk 13 above)
Comment (ezyang) [reply test/test_dynamo_cudagraphs.py:75]: this is a little subtle

First, the context manager is misleading. It doesn't actually turn on optimization for the inside of the manager. Optimization only turns on when you hit a new frame (e.g., do a function call).

With that out of the way, what if you have the optimization inside or outside of a loop? It will depend. If the loop successfully unrolls, then you will get a compiled graph outside the loop that has the unrolled graph. But let's say there's some reason we can't compile the outer frame. Then we will compile the inner function, and the two applications are equivalent/

File: test/test_dynamo_cudagraphs.py (hunk 14 above)
Comment (ezyang) [reply test/test_dynamo_cudagraphs.py:109]: Functionalize gives you a graph that doesn't have mutating operations in it. CUDA graphs actually doesn't want functionalization, but we need it because there are passes like the partitioner we use here which are unsound in the presence of mutation.

File: torch/cuda/_dynamo_graphs.py (hunk 16 above)
Comment (ezyang) [reply torch/cuda/_dynamo_graphs.py:44]: Yes, this is a general pattern for doing compilation on FX graph directly.

File: torch/cuda/_dynamo_graphs.py (hunk 17 above)
Comment (ezyang) [reply torch/cuda/_dynamo_graphs.py:4]: Suppose you have a program like

```
cuda op 1
cuda op 2
cuda op 3
cpu op
cuda op 4
cuda op 5
cuda op 6
```

partitioning puts cuda 1-3 and 4-6 in separate fusion groups that cuda graphs can be applied on (cuda graphs can't be used on cpu ops)

File: torch/cuda/_dynamo_graphs.py (hunk 18 above)
Comment (ezyang) [reply torch/cuda/_dynamo_graphs.py:45]: sort of. But our tracing (proxy tensor) operates at a different level than cuda graph capture. Our tracing capture aten ops; cuda graph captures cuda kernel launches

File: torch/fx/passes/backends/cudagraphs.py (hunk 19 above)
Comment (ezyang) [reply torch/fx/passes/backends/cudagraphs.py:10]: yep
